File: backend/libbackend/account.ml
open Core_kernel
open Libexecution
open Types
module Hash = Sodium.Password_hash.Bytes

let banned_usernames : string list =
  (* originally from https://ldpreload.com/blog/names-to-reserve *)
  (* we allow www, because we have a canvas there *)
  [ "abuse"
  ; "admin"
  ; "administrator"
  ; "autoconfig"
  ; "broadcasthost"
  ; "ftp"
  ; "hostmaster"
  ; "imap"
  ; "info"
  ; "is"
  ; "isatap"
  ; "it"
  ; "localdomain"
  ; "localhost"
  ; "mail"
  ; "mailer-daemon"
  ; "marketing"
  ; "mis"
  ; "news"
  ; "nobody"
  ; "noc"
  ; "noreply"
  ; "no-reply"
  ; "pop"
  ; "pop3"
  ; "postmaster"
  ; "root"
  ; "sales"
  ; "security"
  ; "smtp"
  ; "ssladmin"
  ; "ssladministrator"
  ; "sslwebmaster"
  ; "support"
  ; "sysadmin"
  ; "usenet"
  ; "uucp"
  ; "webmaster"
  ; "wpad" ]
  @ (* original to us *)
    ["billing"; "dev"]


type username = string [@@deriving yojson]

type account =
  { username : username
  ; password : Password.t
  ; email : string
  ; name : string }

type user_info =
  { username : username
  ; email : string
  ; name : string
  ; admin : bool }
[@@deriving yojson]

type user_info_and_created_at =
  { username : username
  ; email : string
  ; name : string
  ; admin : bool
  ; created_at : string }
[@@deriving yojson]

(************************)
(* Adding *)
(************************)
let validate_username (username : string) : (unit, string) Result.t =
  (* rules: no uppercase, ascii only, must start with letter, other letters can
   * be numbers or underscores. 3-20 characters. *)
  let regex = Re2.create_exn "^[a-z][a-z0-9_]{2,20}$" in
  if Re2.matches regex username
  then Ok ()
  else
    Error
      ( "Invalid username '"
      ^ username
      ^ "', must match /^[a-z][a-z0-9_]{2,20}$/" )


let validate_email (email : string) : (unit, string) Result.t =
  (* just checking it's roughly the shape of an email *)
  let regex = Re2.create_exn ".+@.+\\..+" in
  if Re2.matches regex email
  then Ok ()
  else Error ("Invalid email '" ^ email ^ "'")


let validate_account (account : account) : (unit, string) Result.t =
  validate_username account.username
  |> Prelude.Result.and_ (validate_email account.email)


let upsert_account ?(validate : bool = true) (account : account) :
    (unit, string) Result.t =
  let result = if validate then validate_account account else Ok () in
  Result.map result ~f:(fun () ->
      Db.run
        ~name:"upsert_account"
        ~subject:account.username
        "INSERT INTO accounts
    (id, username, name, email, admin, password)
    VALUES
    ($1, $2, $3, $4, false, $5)
    ON CONFLICT (username)
    DO UPDATE SET name = EXCLUDED.name,
                  email = EXCLUDED.email,
                  password = EXCLUDED.password"
        ~params:
          [ Uuid (Util.create_uuid ())
          ; String account.username
          ; String account.name
          ; String account.email
          ; String (Password.to_bytes account.password) ] )


let upsert_account_exn ?(validate : bool = true) (account : account) : unit =
  upsert_account ~validate account
  |> Prelude.Result.ok_or_internal_exception "Cannot upsert account"


let upsert_admin ?(validate : bool = true) (account : account) :
    (unit, string) Result.t =
  Result.map (validate_account account) ~f:(fun () ->
      Db.run
        ~name:"upsert_admin"
        ~subject:account.username
        "INSERT INTO accounts as u
    (id, username, name, email, admin, password)
    VALUES
    ($1, $2, $3, $4, true, $5)
    ON CONFLICT (username)
    DO UPDATE SET name = EXCLUDED.name,
                  email = EXCLUDED.email,
                  admin = true,
                  password = EXCLUDED.password"
        ~params:
          [ Uuid (Util.create_uuid ())
          ; String account.username
          ; String account.name
          ; String account.email
          ; String (Password.to_bytes account.password) ] )


let upsert_admin_exn ?(validate : bool = true) (account : account) : unit =
  upsert_admin ~validate account
  |> Prelude.Result.ok_or_internal_exception "Cannot upsert account"


(************************)
(* Querying *)
(************************)
let username_of_id id =
  Db.fetch_one_option
    ~name:"account_of_id"
    ~subject:(Uuidm.to_string id)
    "SELECT username from accounts
     WHERE accounts.id = $1"
    ~params:[Uuid id]
  |> Option.map ~f:List.hd_exn


let id_of_username username : Uuidm.t option =
  Db.fetch_one_option
    ~name:"account_of_username"
    ~subject:username
    "SELECT id from accounts
     WHERE accounts.username = $1"
    ~params:[String username]
  |> Option.map ~f:List.hd_exn
  |> fun x -> match x with Some sid -> Uuidm.of_string sid | None -> None


let get_user username =
  Db.fetch_one_option
    ~name:"get_user"
    ~subject:username
    "SELECT name, email, admin from accounts
     WHERE accounts.username = $1"
    ~params:[String username]
  |> Option.bind ~f:(function
         | [name; email; admin] ->
             Some {username; name; admin = admin = "t"; email}
         | _ ->
             None )


let get_user_and_created_at username =
  Db.fetch_one_option
    ~name:"get_user_and_created_at"
    ~subject:username
    "SELECT name, email, admin, created_at from accounts
     WHERE accounts.username = $1"
    ~params:[String username]
  |> Option.bind ~f:(function
         | [name; email; admin; created_at] ->
             Some
               { username
               ; name
               ; admin = admin = "t"
               ; email
               ; created_at =
                   created_at
                   |> Db.date_of_sqlstring
                   |> Core.Time.to_string_iso8601_basic
                        ~zone:Core.Time.Zone.utc }
         | _ ->
             None )


let get_user_by_email email =
  Db.fetch_one_option
    ~name:"get_user_by_email"
    ~subject:email
    "SELECT name, username, admin from accounts
     WHERE accounts.email = $1"
    ~params:[String email]
  |> Option.bind ~f:(function
         | [name; username; admin] ->
             Some {username; name; admin = admin = "t"; email}
         | _ ->
             None )


let get_users () =
  Db.fetch ~name:"get_users" "SELECT username from accounts" ~params:[]
  |> List.map ~f:List.hd_exn


let is_admin ~username : bool =
  Db.exists
    ~subject:username
    ~name:"is_admin"
    "SELECT 1 from accounts
     WHERE accounts.username = $1
       AND accounts.admin = true"
    ~params:[String username]


(* Any external calls to this should also call Stroller.segment_identify_user;
 * we can't do it here because that sets up a module dependency cycle *)
let set_admin ~username (admin : bool) : unit =
  Db.run
    ~name:"set_admin"
    ~subject:username
    "UPDATE accounts SET admin = $1 where username = $2"
    ~params:[Bool admin; String username]


(* Returns None if no valid user, or Some username _from the db_ if valid. Note:
 * the input username may also be an email address.
 *
 * No need to detect which and SQL differently; no valid username contains a
 * '@', and every valid email address does. [If you say 'uucp bang path', I will
 * laugh and then tell you to give me a real email address.] *)
let valid_user ~(username : username) ~(password : string) : string option =
  match
    Db.fetch_one_option
      ~name:"valid_user"
      ~subject:username
      "SELECT username, password from accounts
           WHERE accounts.username = $1 OR accounts.email = $1"
      ~params:[String username]
  with
  | Some [db_username; db_password] ->
      if password
         |> Bytes.of_string
         |> Hash.wipe_to_password
         |> Hash.verify_password_hash
              (Bytes.of_string (B64.decode db_password))
      then Some db_username
      else None
  | None | _ ->
      None


let can_access_operations ~(username : username) : bool = is_admin ~username

(* Returns None if no valid user, or Some username _from the db_ if valid. Note:
 * the input username may also be an email address *)
let authenticate ~(username : username) ~(password : string) : string option =
  valid_user ~username ~password


let owner ~(auth_domain : string) : Uuidm.t option =
  let auth_domain = String.lowercase auth_domain in
  if List.mem banned_usernames auth_domain ~equal:( = )
  then None
  else
    Db.fetch_one_option
      ~name:"owner"
      ~subject:auth_domain
      "SELECT id from accounts
     WHERE accounts.username = $1"
      ~params:[String auth_domain]
    |> Option.map ~f:List.hd_exn
    |> Option.bind ~f:Uuidm.of_string


let auth_domain_for host : string =
  match String.split host '-' with d :: _ -> d | _ -> host


let for_host (host : string) : Uuidm.t option =
  host |> auth_domain_for |> owner


let for_host_exn (host : string) : Uuidm.t =
  host
  |> for_host
  |> fun o -> Option.value_exn ~message:("No owner found for host " ^ host) o


(************************)
(* Darkinternal functions *)
(************************)

(* Any external calls to this should also call Stroller.segment_identify_user;
 * we can't do it here because that sets up a module dependency cycle *)
let upsert_user ~(username : string) ~(email : string) ~(name : string) () :
    (string, string) Result.t =
  let plaintext = Util.random_string 16 in
  let password = Password.from_plaintext plaintext in
  upsert_account {username; email; name; password}
  |> Result.map ~f:(fun () -> plaintext)


let init_testing () : unit =
  upsert_account_exn
    { username = "test_unhashed"
    ; password = Password.from_hash "fVm2CUePzGKCwoEQQdNJktUQ"
    ; email = "test+unhashed@darklang.com"
    ; name = "Dark OCaml Tests with Unhashed Password" } ;
  upsert_account_exn
    { username = "test"
    ; password = Password.from_plaintext "fVm2CUePzGKCwoEQQdNJktUQ"
    ; email = "test@darklang.com"
    ; name = "Dark OCaml Tests" } ;
  upsert_admin_exn
    { username = "test_admin"
    ; password = Password.from_plaintext "fVm2CUePzGKCwoEQQdNJktUQ"
    ; email = "test+admin@darklang.com"
    ; name = "Dark OCaml Test Admin" } ;
  ()


let upsert_admins () : unit =
  upsert_admin_exn
    { username = "ian"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkOXd2R3BSYW54Y3llYmdRdU1EMHdUdyRNN1ljWVFQdDk0S29nM1EyM1Q2cHFRZDRlMk9VM3lDTmpreUZ2NGIva1o4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "ian@darklang.com"
    ; name = "Ian Connolly" } ;
  upsert_admin_exn
    { username = "paul"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkcEQxWXBLOG1aVStnUUJUYXdKZytkQSR3TWFXb1hHOER1UzVGd2NDYzRXQVc3RlZGN0VYdVpnMndvZEJ0QnY1bkdJAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "paul@darklang.com"
    ; name = "Paul Biggar" } ;
  upsert_admin_exn
    { username = "ellen"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkcHcxNmRhelJaTGNrYXhZV1psLytXdyRpUHJ1V1NQV2xya1RDZjRDbGlwNTkyaC9tSlZvaTVWSTliRlp0c2xrVmg0AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "ellen@darklang.com"
    ; name = "Ellen Chisa" } ;
  upsert_admin_exn
    { username = "alice"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkVGllNGtJT3kyMVFjL1dGUnhScC9PdyROMnp1ZVZnczhIcjl0ODZEREN2VFBYMVNHOE1Za1plSUZCSWFzck9aR1J3AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "alice@darklang.com"
    ; name = "Alice Wong" } ;
  upsert_admin_exn
    { username = "ismith"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkbHlXamc0MHA3MWRBZ1kyTmFTTVhIZyRnaWZ1UGpsSnoxMFNUVDlZYWR5Tis1SVovRFVxSXdZeXVtL0Z2TkFOa1ZnAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "ismith@darklang.com"
    ; name = "Ian Smith" } ;
  upsert_admin_exn
    { username = "sydney"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkMDJYZzhSS1RQai9JOGppdzI5MTBEUSRJdE0yYnlIK29OL1RIdzFJbC9yNWZBT2RGR0xrUFc3V3MxaVpUUUVFKytjAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "sydney@darklang.com"
    ; name = "Sydney Noteboom" } ;
  upsert_admin_exn
    { username = "korede"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkRGVxb0M0dXJUYkltWWdlYmRidGQxZyRXTHNrRTErTThscmwvRUlIVGoxUFpVVE5nNDdNQ0FqVHZRWHFvMVFjUkI4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "korede@darklang.com"
    ; name = "Korede" } ;
  upsert_admin_exn
    { username = "julian"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkRmFQeGdPZXhaZTFZL2pSYkZ4azFNQSRFR0ZyNEkyeDVqaDIvL243UEIzeUhkcTIwaUZUUi91RXE1RDJkQ0o0eE5BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "julian@darklang.com"
    ; name = "Julian Ceipek" } ;
  upsert_admin_exn
    { username = "dean"
    ; password =
        Password.from_hash
          "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkWjdFdjJlZ2ZmMnZRaFhjQWlpOWlPZyRrL2F1bGFEU0tra3BQMmNKTHF6S3NaU3d5WXdmWm1pNkQ4Yy96alJrT3YwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
    ; email = "dean@darklang.com"
    ; name = "Dean Strelau" } ;
  (* dark contractors *)
  (* upsert_account_exn *)
  (*   { username = "lizzie" *)
  (*   ; password = *)
  (*       "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkWTNueDFWQUFYRWpLMjJGclcwMjU2ZyRYVDQxUGtGNnYyM1E4L0MrSUZITlNXNi8wUGN4TFdEbkRMZ0xVdHN2bHJZAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=" *)
  (*   ; email = "_@lizzie.io" *)
  (*   ; name = "Lizzie Dixon" } ; *)
  ()


(* accounts to create namespaces for dark canvases *)
let upsert_useful_canvases () : unit =
  (* Needed for tests *)
  upsert_admin_exn
    ~validate:false
    { username = "sample"
    ; password = Password.invalid
    ; email = "nouser@example.com"
    ; name = "Sample Owner" }


let upsert_banned_accounts () : unit =
  ignore
    ( banned_usernames
    |> List.map ~f:(fun username ->
           upsert_account_exn
             ~validate:false
             { username
             ; password = Password.invalid
             ; email = "ops+" ^ username ^ "@darklang.com"
             ; name = "Disallowed account" } ) ) ;
  ()


let init () : unit =
  if Config.create_accounts
  then (
    init_testing () ;
    upsert_banned_accounts () ;
    upsert_admins () ;
    upsert_useful_canvases () ;
    () )


module Testing = struct
  let validate_username = validate_username

  let validate_email = validate_email
end

File: fsharp-backend/src/LibBackend/Account.fs
module LibBackend.Account

// Functions related to Accounts/Users

open System.Threading.Tasks
open FSharp.Control.Tasks
open Npgsql.FSharp
open Npgsql

open Prelude
open Tablecloth
open Db

module RT = LibExecution.RuntimeTypes
module DvalRepr = LibExecution.DvalRepr

// **********************
// Types
// **********************

type Account =
  { username : UserName.T
    password : Password.T
    email : string
    name : string }

type UserInfo =
  { username : UserName.T
    name : string
    admin : bool
    email : string
    id : UserID }

type UserInfoAndCreatedAt =
  { username : UserName.T
    name : string
    admin : bool
    email : string
    id : UserID
    createdAt : System.DateTime }

let userInfoToPerson (ui : UserInfo) : LibService.Rollbar.Person =
  { id = ui.id; email = ui.email; username = ui.username }



// **********************
// Special usernames
// **********************



// **********************
// Adding
// **********************


let validateEmail (email : string) : Result<unit, string> =
  (* just checking it's roughly the shape of an email *)
  let reString = "^.+@.+\\..+$"

  if FsRegEx.isMatch reString email then
    Ok()
  else
    Error($"Invalid email '{email}'")


let validateAccount (account : Account) : Result<unit, string> =
  validateEmail account.email

// Passwords set here are only valid locally, production uses auth0 to check
// access
let upsertAccount (admin : bool) (account : Account) : Task<Result<unit, string>> =
  task {
    match validateAccount account with
    | Ok () ->
      return!
        Sql.query
          "INSERT INTO accounts
             (id, username, name, email, admin, password)
             VALUES
             (@id, @username, @name, @email, @admin, @password)
             ON CONFLICT (username)
             DO UPDATE SET name = EXCLUDED.name,
                           email = EXCLUDED.email,
                           password = EXCLUDED.password"
        |> Sql.parameters [ "id", Sql.uuid (System.Guid.NewGuid())
                            "username", Sql.string (string account.username)
                            "admin", Sql.bool admin
                            "name", Sql.string account.name
                            "email", Sql.string account.email
                            ("password", account.password |> string |> Sql.string) ]
        |> Sql.executeStatementAsync
        |> Task.map Ok
    | Error _ as result -> return result
  }

let upsertAdmin = upsertAccount true
let upsertNonAdmin = upsertAccount false

// Any external calls to this should also call Analytics.identifyUser;
// we can't do it here because that sets up a module dependency cycle
let insertUser
  (username : UserName.T)
  (email : string)
  (name : string)
  (analyticsMetadata : Option<RT.DvalMap>)
  : Task<Result<unit, string>> =
  task {
    let account =
      // As of the move to auth0, we  no longer store passwords in postgres. We do
      // still use postgres locally, which is why we're not removing the field
      // entirely. Local account creation is done in
      // upsert_account_exn/upsert_admin_exn, so using Password.invalid here does
      // not affect that
      { username = username
        password = Password.invalid
        email = email
        name = name }

    match validateAccount account with
    | Ok () ->
      let analyticsMetadata = analyticsMetadata |> Option.unwrap (Map.empty)

      try
        do!
          Sql.query
            "INSERT INTO accounts
              (id, username, name, email, admin, password, segment_metadata)
              VALUES
              (@id, @username, @name, @email, false, @password, @metadata)
              ON CONFLICT DO NOTHING"
          |> Sql.parameters [ "id", Sql.uuid (System.Guid.NewGuid())
                              "username", Sql.string (string username)
                              "name", Sql.string name
                              "email", Sql.string email
                              ("password", Sql.string (string Password.invalid))
                              ("analytics_metadata",
                               Sql.jsonb (
                                 DvalRepr.toInternalQueryableV1 analyticsMetadata
                               )) ]
          |> Sql.executeStatementAsync
        let! exists =
          Sql.query
            "SELECT 1 from ACCOUNTS
              WHERE username = @username
                AND name = @name
                AND email = @email
                AND password = @password"
          |> Sql.parameters [ "username", Sql.string (string username)
                              "name", Sql.string name
                              "email", Sql.string email
                              ("password", Sql.string (string Password.invalid)) ]
          |> Sql.executeExistsAsync
        if exists then
          return Ok()
        else
          return
            Error "Insert failed, probably because the username is already taken."
      with
      | _ ->
        return Error "Insert failed, probably because the username is already taken."
    | Error e -> return Error e
  }






// **********************
// Querying
// **********************

let userIDForUserName (username : UserName.T) : Task<UserID> =
  Sql.query
    "SELECT id
       FROM accounts
       WHERE accounts.username = @username"
  |> Sql.parameters [ "username", Sql.string (string username) ]
  |> Sql.executeRowOptionAsync (fun read -> read.uuid "id")
  |> Task.map (function
    | Some v -> v
    | None -> Exception.raiseDeveloper "User not found")

let usernameForUserID (userID : UserID) : Task<Option<UserName.T>> =
  Sql.query
    "SELECT username
     FROM accounts
     WHERE accounts.id = @userid"
  |> Sql.parameters [ "userid", Sql.uuid userID ]
  |> Sql.executeRowOptionAsync (fun read -> read.string "username" |> UserName.create)

let getUser (username : UserName.T) : Task<Option<UserInfo>> =
  Sql.query
    "SELECT name, email, admin, id
     FROM accounts
     WHERE accounts.username = @username"
  |> Sql.parameters [ "username", Sql.string (string username) ]
  |> Sql.executeRowOptionAsync (fun read ->
    { username = username
      name = read.string "name"
      email = read.string "email"
      admin = read.bool "admin"
      id = read.uuid "id" })

let getUserCreatedAt (username : UserName.T) : Task<System.DateTime> =
  Sql.query
    "SELECT created_at
     FROM accounts
     WHERE accounts.username = @username"
  |> Sql.parameters [ "username", Sql.string (string username) ]
  |> Sql.executeRowAsync (fun read -> read.dateTime "created_at")

let getUserAndCreatedAtAndAnalyticsMetadata
  (username : UserName.T)
  : Task<Option<UserInfoAndCreatedAt * string>> =
  Sql.query
    "SELECT name, email, admin, created_at, id, segment_metadata
     FROM accounts
     WHERE accounts.username = @username"
  |> Sql.parameters [ "username", Sql.string (string username) ]
  |> Sql.executeRowOptionAsync (fun read ->
    { username = username
      name = read.string "name"
      email = read.string "email"
      admin = read.bool "admin"
      id = read.uuid "id"
      createdAt = read.dateTime "created_at" },
    read.string "segment_metadata")

let getUserByEmail (email : string) : Task<Option<UserInfo>> =
  Sql.query
    "SELECT name, username, admin, id
     FROM accounts
     WHERE accounts.email = @email"
  |> Sql.parameters [ "email", Sql.string email ]
  |> Sql.executeRowOptionAsync (fun read ->
    { username = UserName.create (read.string "username")
      name = read.string "name"
      email = email
      admin = read.bool "admin"
      id = read.uuid "id" })

let getUsers () : Task<List<UserName.T>> =
  Sql.query
    "SELECT username
     FROM accounts"
  |> Sql.executeAsync (fun read -> UserName.create (read.string "username"))

let isAdmin (username : UserName.T) : Task<bool> =
  Sql.query
    "SELECT TRUE
     FROM accounts
     WHERE accounts.username = @username
       AND admin = true"
  |> Sql.parameters [ "username", Sql.string (string username) ]
  |> Sql.executeExistsAsync

let setAdmin (admin : bool) (username : UserName.T) : Task<unit> =
  Sql.query
    "UPDATE accounts
        SET admin = @admin where username = @username"
  |> Sql.parameters [ "admin", Sql.bool admin
                      "username", Sql.string (string username) ]
  |> Sql.executeStatementAsync

// Returns None if no valid user, or Some username _from the db_ if valid.
// Note: the input username may also be an email address. We do this because
// users input data this way and it seems silly not to allow it.
//
// No need to detect which and SQL differently; no valid username contains a
// '@', and every valid email address does. [If you say 'uucp bang path', I
// will laugh and then tell you to give me a real email address.]
//
// This function was converted from OCaml. The OCaml Libsodium
// (https://github.com/ahrefs/ocaml-sodium/blob/master/lib/sodium.ml), the F#
// version is libsodium-net
// (https://github.com/tabrath/libsodium-core/blob/master/src/Sodium.Core/PasswordHash.cs).
// The OCaml version uses the argon2i versions under the hood, which we use explicitly in F#.
let authenticate
  (usernameOrEmail : string)
  (givenPassword : string)
  : Task<Option<string>> =
  Sql.query
    "SELECT username, password from accounts
      WHERE accounts.username = @usernameOrEmail OR accounts.email = @usernameOrEmail"
  |> Sql.parameters [ "usernameOrEmail", Sql.string usernameOrEmail ]
  |> Sql.executeRowOptionAsync (fun read ->
    (read.string "username", read.string "password"))
  |> Task.map (
    Option.andThen (fun (username, password) ->
      let dbHash = password |> Base64.decodeFromString |> UTF8.ofBytesWithReplacement

      if Sodium.PasswordHash.ArgonHashStringVerify(dbHash, givenPassword) then
        Some(username)
      else
        None)
  )

let canAccessOperations (username : UserName.T) : Task<bool> = isAdmin username

// formerly called auth_domain_for
let ownerNameFromCanvasName (host : CanvasName.T) : OwnerName.T =
  match String.split "-" (string host) with
  | owner :: _ -> OwnerName.create owner
  | _ -> OwnerName.create (string host)

// **********************
// What user has access to
// **********************

let ownedCanvases (userID : UserID) : Task<List<CanvasName.T>> =
  Sql.query
    "SELECT DISTINCT c.name
     FROM canvases c
     WHERE c.account_id = @userID"
  |> Sql.parameters [ "userID", Sql.uuid userID ]
  |> Sql.executeAsync (fun read -> read.string "name" |> CanvasName.create)
  |> Task.map List.sort


// NB: this returns canvases an account has access to via an organization, not
// the organization(s) themselves
let accessibleCanvases (userID : UserID) : Task<List<CanvasName.T>> =
  Sql.query
    "SELECT c.name
       FROM access
      INNER JOIN accounts as org on access.organization_account = org.id
      INNER JOIN canvases as c on org.id = account_id
      WHERE access.access_account = @userID"
  |> Sql.parameters [ "userID", Sql.uuid userID ]
  |> Sql.executeAsync (fun read -> read.string "name" |> CanvasName.create)
  |> Task.map List.sort

let orgs (userID : UserID) : Task<List<OrgName.T>> =
  Sql.query
    "SELECT org.username
     FROM access
     INNER JOIN accounts as org on access.organization_account = org.id
     WHERE access.access_account = @userID"
  |> Sql.parameters [ "userID", Sql.uuid userID ]
  |> Sql.executeAsync (fun read -> read.string "username" |> OrgName.create)
  |> Task.map List.sort


// **********************
// Local/test developement
// **********************

let initTestAccounts () : Task<unit> =
  task {
    do!
      upsertNonAdmin
        { username = UserName.create "test_unhashed"
          password = Password.fromHash "fVm2CUePzGKCwoEQQdNJktUQ"
          email = "test+unhashed@darklang.com"
          name = "Dark OCaml Tests with Unhashed Password" }
      |> Task.map (Exception.unwrapResultInternal [])

    do!
      upsertNonAdmin
        { username = UserName.create "test"
          password = Password.fromPlaintext "fVm2CUePzGKCwoEQQdNJktUQ"
          email = "test@darklang.com"
          name = "Dark OCaml Tests" }
      |> Task.map (Exception.unwrapResultInternal [])

    do!
      upsertAdmin
        { username = UserName.create "test_admin"
          password = Password.fromPlaintext "fVm2CUePzGKCwoEQQdNJktUQ"
          email = "test+admin@darklang.com"
          name = "Dark OCaml Test Admin" }
      |> Task.map (Exception.unwrapResultInternal [])

    return ()
  }

let initBannedAccounts () : Task<unit> =
  task {
    do!
      UserName.banned
      |> Set.toList
      |> Task.iterSequentially (fun username ->
        upsertNonAdmin
          { username = UserName.createUnsafe username
            password = Password.invalid
            email = $"ops+{username}@darklang.com"
            name = $"Disallowed account {username}" }
        |> Task.map (Exception.unwrapResultInternal []))

    return ()
  }

let initAdmins () : Task<unit> =
  task {
    let password =
      // "what"
      Password.fromHash
        "JGFyZ29uMmkkdj0xOSRtPTMyNzY4LHQ9NCxwPTEkcEQxWXBLOG1aVStnUUJUYXdKZytkQSR3TWFXb1hHOER1UzVGd2NDYzRXQVc3RlZGN0VYdVpnMndvZEJ0QnY1bkdJAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="

    do!
      upsertAdmin
        { username = UserName.create "dark"
          password = password
          email = "ops+darkuser@darklang.com"
          name = "Dark Local Admin user" }
      |> Task.map (Exception.unwrapResultInternal [])

    do!
      upsertAdmin
        { username = UserName.create "paul"
          password = password
          email = "paul@darklang.com"
          name = "Paul Biggar" }
      |> Task.map (Exception.unwrapResultInternal [])

    return ()
  }

// accounts to create namespaces for dark canvases
let initUsefulCanvases () : Task<unit> =
  // Needed for tests
  task {
    do!
      upsertNonAdmin
        { username = UserName.create "sample"
          password = Password.invalid
          email = "opsample@darklang.com"
          name = "Sample owner" }
      |> Task.map (Exception.unwrapResultInternal [])

    return ()
  }


let init (serviceName : string) : Task<unit> =
  task {
    print $"Initing LibBackend in {serviceName}"
    if Config.createAccounts then
      do! initTestAccounts ()
      do! initBannedAccounts ()
      do! initAdmins ()
      do! initUsefulCanvases ()
      return ()
    else
      return ()
  }
